### PR TITLE
Add `face_to_cell_index_nodal` for simplex elements

### PR DESCRIPTION
**Pull request by dominiktassilostill**
In combination with #17205 this is useful to find the DoF index of the cell when operating on the integration points on a face, e.g. one can evaluate a boundary function at the location of the DoF and then insert it into the corresponding entry of the DoF vector.
To test it I compared it with the row having the non-zero entry for the corresponding integration point in `shape_values_face`.

### CODE REVIEW

File: include/deal.II/matrix_free/shape_info.templates.h [hunk 1]
```diff
@@ -496,6 +496,96 @@ namespace internal
...
+                           d < dofs_per_component_on_face - 1;
+                           ++line, d += 2)
+                        {
+                          // check the direction of the line
+                          // is it 0 -> 1 or 1 -> 0
+                          // as DoFs on the line are ordered differently
+                          if (reference_cell.line_to_cell_vertices(
+                                reference_cell.face_to_cell_lines(face,
+                                                                  line,
+                                                                  1),
```
Comment (kronbichler) [inline include/deal.II/matrix_free/shape_info.templates.h:544]: If I read the code correctly, you're using `reference_cell.face_to_cell_lines` multiple times with the same arguments. The additional line breaks in the formula here and below are also not helpful. I would suggest to create a local variable, say `const unsigned int face_to_cell_lines`, and use that in all appearances.

**Review by kronbichler — COMMENTED** [review]
I think this looks good for the present code. Could you please add a short test that simply prints these numbers, out of a generated `ShapeInfo` for the continuous and discontinuous simplex elements in 2D and 3D?

File: include/deal.II/matrix_free/shape_info.templates.h [hunk 2]
```diff
@@ -496,6 +496,96 @@ namespace internal
...
+
+
+              for (unsigned int face = 0; face < reference_cell.n_faces();
+                   ++face)
+                {
+                  // first get info from reference cell, i.e. the linear case
+                  unsigned int d = 0;
+                  for (; d < dim; ++d)
+                    face_to_cell_index_nodal[face][d] =
+                      reference_cell.face_to_cell_vertices(face, d, 1);
```
[the remark this answers is not in the document]
Comment (dominiktassilostill) [reply include/deal.II/matrix_free/shape_info.templates.h:519]: The CG and DG element share the same Lagrange basis and unit support points, so as far as I understand both should follow the same numbering. Maybe @peterrum can confirm this?

File: include/deal.II/matrix_free/shape_info.templates.h (hunk 1 above)
Comment (dominiktassilostill) [reply include/deal.II/matrix_free/shape_info.templates.h]: Very good suggestion, it is much easier to read now!

File: include/deal.II/matrix_free/shape_info.templates.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (peterrum) [reply include/deal.II/matrix_free/shape_info.templates.h:519]: > Maybe @peterrum can confirm this?

For now, yes. But maybe there is something more reasonable for DG, similarly to the lexicographical ordering as in the FE_DGQ case.

File: tests/matrix_free/face_to_cell_index_nodal_simplex.cc [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+template <int dim>
+void
+test(const FiniteElement<dim> &fe, const Quadrature<dim> &quad)
+{
+  for (unsigned int i = 0; i < fe.n_base_elements(); ++i)
+    {
+      internal::MatrixFreeFunctions::ShapeInfo<double> shape_info;
+      shape_info.reinit(quad, fe, i);
+      deallog << "Dim: " << dim << ", degree: " << fe.degree << std::endl;
```
Comment (kronbichler) [inline tests/matrix_free/face_to_cell_index_nodal_simplex.cc:43]:
```suggestion
      deallog << "Testing: " << fe.get_name() << std::endl;
```
This should print all information, both the dimension, degree and dG/cG

**Comment by peterrum** [comment]
Is there a reason why not to use `DoFHandler`/`DoFAccessor`? `DoFAccessor` is recursively implementing what you have done:

```cpp
#include <deal.II/dofs/dof_handler.h>

#include <deal.II/fe/fe_simplex_p.h>

#include <deal.II/grid/grid_generator.h>

using namespace dealii;

int
main()
{
  const unsigned int dim       = 3;
  const unsigned int fe_degree = 3;

  Triangulation<dim> tria;
  GridGenerator::reference_cell(tria,
                                ReferenceCells::template get_simplex<dim>());

  FE_SimplexP<dim> fe(fe_degree);

  DoFHandler<dim> dof_handler(tria);
  dof_handler.distribute_dofs(fe);

  std::vector<types::global_dof_index> dof_indices(fe.n_dofs_per_face());

  for (const auto &cell : dof_handler.active_cell_iterators())
    for (const auto &face : cell->face_iterators())
      {
        face->get_dof_indices(dof_indices);

        for (const auto i : dof_indices)
          std::cout << i << " ";
        std::cout << std::endl;
      }
}

```

There are other places where we set up a dummy `Triangulation`/`DoFHandler` with a single cell.

**Comment by kronbichler** [comment]
> Is there a reason why not to use DoFHandler/DoFAccessor? [...] There are other places where we set up a dummy Triangulation/DoFHandler with a single cell.

`ShapeInfo` is a function that is called relatively often and in a number of places, so I would hesitate to set up a whole `DoFHandler` and `Triangulation` object just to extract the numbers. We might have inefficiencies in the tetrahedral code path so the overall cost could be ok, but for hexahedral elements I would be strongly opposed to such an implementation in the long run. In either case, I think the present code will be some intermediate solution before we manage to complete #9544, so I do not feel strongly at this point, if we commit ourselves to soon address that issue.

**Comment by dominiktassilostill** [comment]
> Is there a reason why not to use DoFHandler/DoFAccessor?

Nothing in particular apart from what @kronbichler said, your implementation would have the advantage of also working with Pyramids and Wedges correct? So maybe this would be an option for these 2?

**Comment by peterrum** [comment]
I am fine with keeping it as it is now. Could we however change the test so that the indices are compared to the ones from `DoFHandler` instead of printing them to the screen.

> your implementation would have the advantage of also working with Pyramids and Wedges correct? So maybe this would be an option for these 2?

Yes it would.

**Review by kronbichler — APPROVED** [review]

**Pull request merged**